kvserver: rework memory allocation in replicastats

kvoli opened the pull request:
This patch removes some unused fields within the replica stats object.
It also opts to allocate all the memory needed upfront for a replica
stats object for better cache locality and less GC overhead.

This patch also removes locality tracking for the other throughput trackers
to reduce per-replica memory footprint.

resolves #85112

Release justification: low risk, lowers memory footprint to avoid oom.
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86592)
<!-- Reviewable:end -->

irfansharif reviewed (commented):
Reducing the size of the struct like we've done is good, but the total count of objects allocated is important which can cause undue lot of pressure on the Go GC. It seems that for every *ReplicaStats object we have 6 underlying *replicaStatsRecord objects.

https://github.com/cockroachdb/cockroach/blob/8b03aeacf91ee756cf316d967b5276e9107a069f/pkg/kv/kvserver/replicastats/replica_stats.go#L145-L147

Also for every *Replica on a store, we have 6 *ReplicaStats objects.

https://github.com/cockroachdb/cockroach/blob/afe67ec26f220481264b0279669827e319faa82c/pkg/kv/kvserver/replica_load.go#L36-L43

So that's a multiplier of 6*6=36 objects per replica. Could we get this down to just one object allocated per-replica? I suspect that's the bottleneck more than the size of each object itself.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-:-NAGOYYh0s8-PmpF-OjA:b-u4t824)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lidorcarmel)



<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (changes requested):
> Reviewed 2 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86592)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`pkg/kv/kvserver/replicastats/replica_stats.go` line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-NAHAI2kBefr8TSKl40e:-NAHAI2kBefr8TSKl40f:bgxj8i8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b03aeacf91ee756cf316d967b5276e9107a069f/pkg/kv/kvserver/replicastats/replica_stats.go#L104)):*
> ```Go
> 
> // mergeReplicaStatsRecords combines two records and returns a new record with
> // the merged data. When this is called with nil records, a nil record is
> ```

update the comment? at least replace nil with "inactive" or something like that.
___
*[`pkg/kv/kvserver/replicastats/replica_stats.go` line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-NAHBoRT58Mp1eISgtu5:-NAHBoRT58Mp1eISgtu6:bck1z3s) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b03aeacf91ee756cf316d967b5276e9107a069f/pkg/kv/kvserver/replicastats/replica_stats.go#L107)):*
> ```Go
> // returned; otherwise a new record instead.
> func (rsr *replicaStatsRecord) mergeReplicaStatsRecords(other *replicaStatsRecord) {
> 	if !rsr.active && !other.active {
> ```

can this ever happen? I think no but maybe I'm missing something.
___
*[`pkg/kv/kvserver/replicastats/replica_stats.go` line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-NAHDaCpAxahaFA4SoaM:-NAHDaCpAxahaFA4SoaN:b4pi102) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b03aeacf91ee756cf316d967b5276e9107a069f/pkg/kv/kvserver/replicastats/replica_stats.go#L115)):*
> ```Go
> 	}
> 
> 	if !other.active {
> ```

I think these 2 if statements are also not needed? because we will always reset when deactivating?



_Code quote:_
```Go
	if !rsr.active {
		rsr.reset()







	}

	if !other.active {
		other.reset()
	}
```
___
*[`pkg/kv/kvserver/replicastats/replica_stats.go` line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-NAHDxHEBHz1DmyY_4wF:-NAHDxHEBHz1DmyY_4wG:bfv6dza) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b03aeacf91ee756cf316d967b5276e9107a069f/pkg/kv/kvserver/replicastats/replica_stats.go#L120)):*
> ```Go
> 
> 	rsr.sum += other.sum
> 	rsr.active = true
> ```

btw in the old behavior we "activated" other.. right? so this is different. this new behavior makes more sense to me though.
___
*[`pkg/kv/kvserver/replicastats/replica_stats.go` line 210 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-NAHBwma1-0pgv5b0kFE:-NAHBwma1-0pgv5b0kFF:b3raskw) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b03aeacf91ee756cf316d967b5276e9107a069f/pkg/kv/kvserver/replicastats/replica_stats.go#L210)):*
> ```Go
> 		if !rs.Mu.records[i].active {
> 			other.Mu.records[i].reset()
> 			other.Mu.records[i].active = false
> ```

reset() will already set active to false, right? so maybe drop this line?

can we simplify things by having `activate()` and `deactivate()` or something like that? I feel like I need to track `reset()` together with `active` which is error prone.


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
I agree we should reconsider the structure - however I'm opting to not move to this single object model yet. I don't think it would address the OOM issue that motivated this change and I would rather put in a smaller fix here.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-:-NAKdb398Cgeh4ODZxXC:b5a2yub)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @lidorcarmel)

___
*[`pkg/kv/kvserver/replicastats/replica_stats.go` line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-NAHAI2kBefr8TSKl40e:-NAKfafO1Bz30c_MkfeB:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b03aeacf91ee756cf316d967b5276e9107a069f/pkg/kv/kvserver/replicastats/replica_stats.go#L104)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

update the comment? at least replace nil with "inactive" or something like that.
</blockquote></details>

Updated.
___
*[`pkg/kv/kvserver/replicastats/replica_stats.go` line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-NAHBoRT58Mp1eISgtu5:-NAKFaa2A8c8Yo26eM_y:bt0qivp) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b03aeacf91ee756cf316d967b5276e9107a069f/pkg/kv/kvserver/replicastats/replica_stats.go#L107)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

can this ever happen? I think no but maybe I'm missing something.
</blockquote></details>

This can happen - it's possible that either has only gone through < 6 windows, in which case they would be inactive.

The logic is bloated however,  I've updated to make this clearer.
___
*[`pkg/kv/kvserver/replicastats/replica_stats.go` line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-NAHDaCpAxahaFA4SoaM:-NAKHPbq8STlCp8wOEoa:be55suz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b03aeacf91ee756cf316d967b5276e9107a069f/pkg/kv/kvserver/replicastats/replica_stats.go#L115)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

I think these 2 if statements are also not needed? because we will always reset when deactivating?
</blockquote></details>

Updated - see above.
___
*[`pkg/kv/kvserver/replicastats/replica_stats.go` line 210 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-NAHBwma1-0pgv5b0kFE:-NAKHnMk3zmOFmFFEWki:bsr91zu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b03aeacf91ee756cf316d967b5276e9107a069f/pkg/kv/kvserver/replicastats/replica_stats.go#L210)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

reset() will already set active to false, right? so maybe drop this line?

can we simplify things by having `activate()` and `deactivate()` or something like that? I feel like I need to track `reset()` together with `active` which is error prone.
</blockquote></details>

Good idea. Updated.


<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (approved):
consider filing a short issue about this so we don't forget (unless you can send the pr soon).

> Reviewed 3 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-:-NAKlFtH4wtwJQYMdcrM:b-r88vy5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @kvoli)

___
*[`pkg/kv/kvserver/replicastats/replica_stats.go` line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-NAHBoRT58Mp1eISgtu5:-NAKkzDvEFfdwyDV6vxd:bd5l6mr) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b03aeacf91ee756cf316d967b5276e9107a069f/pkg/kv/kvserver/replicastats/replica_stats.go#L107)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

This can happen - it's possible that either has only gone through < 6 windows, in which case they would be inactive.

The logic is bloated however,  I've updated to make this clearer.
</blockquote></details>

thanks!
___
*[`pkg/kv/kvserver/replicastats/replica_stats.go` line 78 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-NAKiEBcAKigXI0CFjYL:-NAKiEBd4SF6WnexYHHX:b-gv75ky) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c0639eb3fec74ba7d85f3a44b3e70b4e8b65e1/pkg/kv/kvserver/replicastats/replica_stats.go#L78)):*
> ```Go
> 		syncutil.Mutex
> 		idx int
> 		// the window records are initialized once, then each window is reused
> ```

s/the/The/ :)
and a double space on the next line (I missed those!)

thanks for the comment btw.



_Code quote:_
```Go
the
```
___
*[`pkg/kv/kvserver/replicastats/replica_stats.go` line 103 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-NAKk67d0yxzXG_qD7Za:-NAKk67d0yxzXG_qD7Zb:b-7m74ox) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c0639eb3fec74ba7d85f3a44b3e70b4e8b65e1/pkg/kv/kvserver/replicastats/replica_stats.go#L103)):*
> ```Go
> 	// values to avoid churning memory. We could also delete elements that
> 	// haven't been seen in a while, however it makes no difference as the
> 	// memory wouldn't be released (https://github.com/golang/go/issues/20135).
> ```

nice!



_Code quote:_
```Go
	// haven't been seen in a while, however it makes no difference as the
	// memory wouldn't be released (https://github.com/golang/go/issues/20135).
	// To protect against a memory leak, where many unused localities
	// accumulate, create a new map if the number of unused localities was
	// greater than 1.
```
___
*[`pkg/kv/kvserver/replicastats/replica_stats.go` line 240 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-NAKkX_cBbHxi3Jo-_CH:-NAKkX_cBbHxi3Jo-_CI:b-i90kk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c0639eb3fec74ba7d85f3a44b3e70b4e8b65e1/pkg/kv/kvserver/replicastats/replica_stats.go#L240)):*
> ```Go
> 			continue
> 		}
> 		// Otherwise, active the rhs record and split the request count between
> ```

s/active/activate/ probably?



_Code quote:_
```Go
active 
```


<!-- Sent from Reviewable.io -->

kvoli committed:
kvserver: rework memory allocation in replicastats

This patch removes some unused fields within the replica stats object.
It also opts to allocate all the memory needed upfront for a replica
stats object for better cache locality and less GC overhead.

resolves #85112

Release justification: low risk, lowers memory footprint to avoid oom.
Release note: None

kvoli committed:
kvserver: only record request locality for qps

Previously, every replica throughput tracker in `replica_load` would
track the request locality. Only batch requests used this locality for
follow-the-workload leaseholder transfers. This patch removes locality
tracking for the other throughput trackers to reduce per-replica memory
footprint.

Release justification: Low risk, high benefit change to reduce memory
overhead in multi-region clusters.
Release note: None

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86592)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @kvoli, and @lidorcarmel)

___
*[`pkg/kv/kvserver/replicastats/replica_stats.go` line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-NAHBoRT58Mp1eISgtu5:-NAo_sciEKwvpRDMsN_v:bvnrpcm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b03aeacf91ee756cf316d967b5276e9107a069f/pkg/kv/kvserver/replicastats/replica_stats.go#L107)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

thanks!
</blockquote></details>

updated
___
*[`pkg/kv/kvserver/replicastats/replica_stats.go` line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-NAHDxHEBHz1DmyY_4wF:-NAo_tmy6CX8krMEnmUe:b5pbakx) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b03aeacf91ee756cf316d967b5276e9107a069f/pkg/kv/kvserver/replicastats/replica_stats.go#L120)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

btw in the old behavior we "activated" other.. right? so this is different. this new behavior makes more sense to me though.
</blockquote></details>

yeah - changed to be more logical.
___
*[`pkg/kv/kvserver/replicastats/replica_stats.go` line 78 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-NAKiEBcAKigXI0CFjYL:-NAo_rU6CIObbqHWjfDZ:b-qssebl) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c0639eb3fec74ba7d85f3a44b3e70b4e8b65e1/pkg/kv/kvserver/replicastats/replica_stats.go#L78)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

s/the/The/ :)
and a double space on the next line (I missed those!)

thanks for the comment btw.
</blockquote></details>

fixed.
___
*[`pkg/kv/kvserver/replicastats/replica_stats.go` line 240 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86592#-NAKkX_cBbHxi3Jo-_CH:-NAo_yS5BdLFVvqdgXKy:b-cypxrc) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c0639eb3fec74ba7d85f3a44b3e70b4e8b65e1/pkg/kv/kvserver/replicastats/replica_stats.go#L240)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

s/active/activate/ probably?
</blockquote></details>

updated.


<!-- Sent from Reviewable.io -->

kvoli commented:
filed https://github.com/cockroachdb/cockroach/issues/87187 for ref.

Tested w/ 300k ranges.

Minimal (-500mb) difference. Multi-region would show greater distinction - likely O(localities).

Also pushed ~3.48% more throughput on kv95, however that's within a error imo.

```
# preceding commit
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
 1800.0s        0       34185125        18991.5    215.7     92.3    906.0   1744.8  10737.4

# change
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
 1800.0s        0       35372909        19651.4    208.5     92.3    872.4   1677.7   9663.7
```

kvoli commented:
bors r+
